File: src/main/java/me/lokka30/levelledmobs/util/MythicMobUtils.java
package me.lokka30.levelledmobs.util;

import java.lang.reflect.InvocationTargetException;
import java.util.Optional;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.nametag.Definitions;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.misc.MythicMobsMobInfo;
import org.bukkit.Bukkit;
import org.bukkit.plugin.Plugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Used for detecting mobs using Mythic Mobs
 *
 * @author stumper66
 * @since 3.6.0
 */
public class MythicMobUtils {
    public @Nullable static MythicMobsMobInfo getMythicMobInfo(final @NotNull LivingEntityWrapper lmEntity) {
        // the below code was written against MythicMobs v5.0.4-f1007ca3

        /*
           1. The main class of the plugin, MythicBukkit - get field: private MobExecutor mobManager;
           2. from MobExecutor class, call this function while passing the UUID from the mob: public Optional<ActiveMob> getActiveMob(UUID uuid);
           3. in the ActiveMob class, get this field: private transient MythicMob type;
           4. MythicMob is an interface, the class implementation is MobType
           5. the MobType class has the properties we want including:
              private Boolean preventRandomEquipment = Boolean.valueOf(false);
              private Boolean preventOtherDrops = Boolean.valueOf(false);
              private String internalName;
        */

        // io.lumine.mythic.bukkit.MythicBukkit
        final Plugin mmMain = Bukkit.getPluginManager().getPlugin("MythicMobs");
        if (mmMain == null) {
            return null;
        }

        final Definitions def = LevelledMobs.getInstance().getDefinitions();

        try {
            final Object mobExecutorObj = def.field_MM_mobManager.get(mmMain);

            //     public Optional<ActiveMob> getActiveMob(UUID uuid) {
            //       return ((MobRegistry)this.mobRegistry.get()).getActiveMob(uuid); }

            // Optional<io.lumine.mythic.core.mobs.ActiveMob>
            final Optional<?> activeMobObj = (Optional<?>) def.method_MM_getActiveMob.invoke(
                    mobExecutorObj, lmEntity.getLivingEntity().getUniqueId());

            if (activeMobObj.isEmpty()) {
                return null;
            }

            final Object mobTypeObj = def.field_MM_type.get(activeMobObj.get());
            final MythicMobsMobInfo result = new MythicMobsMobInfo();
            result.preventOtherDrops = (boolean) def.field_MM_preventOtherDrops.get(mobTypeObj);
            result.preventRandomEquipment = (boolean) def.field_MM_preventRandomEquipment.get(mobTypeObj);
            result.internalName = (String) def.field_MM_internalName.get(mobTypeObj);

            return result;
        } catch (InvocationTargetException | IllegalAccessException e) {
            Utils.logger.warning("Error getting MythicMob info: " + e.getMessage());
        }
        return null;
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/util/MythicMobUtils.kt
package io.github.arcaneplugins.levelledmobs.util

import java.lang.reflect.InvocationTargetException
import java.util.Optional
import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.result.MythicMobsMobInfo
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import org.bukkit.Bukkit

/**
 * Used for detecting mobs using Mythic Mobs
 *
 * @author stumper66
 * @since 3.6.0
 */
object MythicMobUtils {
    fun getMythicMobInfo(lmEntity: LivingEntityWrapper): MythicMobsMobInfo? {
        // the below code was written against MythicMobs v5.0.4-f1007ca3

        /*
           1. The main class of the plugin, MythicBukkit - get field: private MobExecutor mobManager;
           2. from MobExecutor class, call this function while passing the UUID from the mob: public Optional<ActiveMob> getActiveMob(UUID uuid);
           3. in the ActiveMob class, get this field: private transient MythicMob type;
           4. MythicMob is an interface, the class implementation is MobType
           5. the MobType class has the properties we want including:
              private Boolean preventRandomEquipment = Boolean.valueOf(false);
              private Boolean preventOtherDrops = Boolean.valueOf(false);
              private String internalName;
        */

        // io.lumine.mythic.bukkit.MythicBukkit

        val mmMain = Bukkit.getPluginManager().getPlugin("MythicMobs")
        if (mmMain == null || !mmMain.isEnabled) {
            return null
        }

        val def = LevelledMobs.instance.definitions

        if (def.fieldMMmobManager == null) {
            Log.war("Mythic Mobs is installed but fieldMMmobManager is null")
            return null
        }

        try {
            val mobExecutorObj = def.fieldMMmobManager!![mmMain]

            //     public Optional<ActiveMob> getActiveMob(UUID uuid) {
            //       return ((MobRegistry)this.mobRegistry.get()).getActiveMob(uuid); }

            // Optional<io.lumine.mythic.core.mobs.ActiveMob>
            val activeMobObj = def.methodMMgetActiveMob!!.invoke(
                mobExecutorObj, lmEntity.livingEntity.uniqueId
            ) as Optional<*>

            if (activeMobObj.isEmpty) {
                return null
            }

            val mobTypeObj = def.fieldMMtype!![activeMobObj.get()]
            val result = MythicMobsMobInfo()
            result.preventOtherDrops = def.fieldMMpreventOtherDrops!![mobTypeObj] as Boolean
            result.preventRandomEquipment = def.fieldMMpreventRandomEquipment!![mobTypeObj] as Boolean
            result.internalName = def.fieldMMinternalName!![mobTypeObj] as String

            return result
        } catch (e: InvocationTargetException) {
            Log.war("Error getting MythicMob info: " + e.message)
        } catch (e: IllegalAccessException) {
            Log.war("Error getting MythicMob info: " + e.message)
        }
        return null
    }
}